### ISSUE TITLE
The default quantization: 4bit is ignored by six trainers (distill, classifier/reranker/cross_encoder, prm, moe_lora_routing, unlearn, asr) while the VRAM pre-flight budgets them as QLoRA

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#795]
## The defect

`training.quantization` defaults to `4bit`, and the schema accepts it for every task. Six trainer wrappers never read the field. They load the base at checkpoint precision, and the run gets no warning:

| Trainer | Model load | Reads `quantization`? | Notes |

|---|---|---|---|

| `trainer/distill.py` | `:269` student, `:316` teacher: `from_pretrained(base, trust_remote_code=..., device_map=...)` | no | The frozen teacher also loads unquantised. |

| `trainer/classifier.py` (classifier / reranker / cross_encoder) | `:248` `AutoModelForSequenceClassification.from_pretrained(base, num_labels=..., problem_type=..., trust_remote_code=...)` | no | |

| `trainer/prm.py` | `:265` `from_pretrained(base, torch_dtype=bfloat16 if cuda else float32)` | no | Also never reads `lora.*`: every base parameter trains, so a PRM run is always a full fine-tune. |

| `trainer/mole_routing.py` | `:266` `from_pretrained(base, torch_dtype=bfloat16 if cuda else float32)` | no | Base is frozen; only the gate trains. |

| `trainer/unlearn.py` | `:181` and `:196` `load_model_and_tokenizer(cfg.base, device=..., trust_remote_code=...)` | no | `load_model_and_tokenizer` has a `quantization=` parameter (`utils/live_eval.py:161`) that is not passed, for the policy or for the NPO/RMU reference copy. |

| `trainer/asr.py` | `:221` `WhisperForConditionalGeneration.from_pretrained(base, trust_remote_code=...)` | no | Full fine-tune unless `asr_lora: true`. |

`grep -n quantiz` returns nothing in any of the six files.

## Evidence: accepted at parse time

```

distill           accepted  quantization='4bit'

classifier        accepted  quantization='4bit'

reranker          accepted  quantization='4bit'

cross_encoder     accepted  quantization='4bit'

prm               accepted  quantization='4bit'

moe_lora_routing  accepted  quantization='4bit'

unlearn           accepted  quantization='4bit'

asr               accepted  quantization='4bit'

```

An explicit `quantization: 8bit` is accepted too (checked for `prm` and `distill`).

## Evidence: the pre-flight predicts the wrong run

`commands/train.py:_build_hardware_fit_input` maps `quantization == "4bit"` to `peft = "qlora"` (`:57-63`) for every task, and `_hardware_fit_preflight` (`:106`) has no task filter. For a `task: prm` config on `Qwen/Qwen2.5-1.5B-Instruct` with `batch_size: 4`, on an 8 GB card:

```

prm pre-flight input: HardwareFitInput(params_b=1.5, seq_len=2048, batch_size=4, optimizer='adamw_torch', quant='4bit', peft='qlora', gradient_checkpointing=False)

  as peft=qlora quant=4bit: ok=True breakdown=VRAMBreakdown(weights_gb=0.75, optimizer_gb=0.12, gradients_gb=0.06, activations_gb=2.5381983657704925, overhead_gb=0.5375)

  as peft=full quant=none: ok=False breakdown=VRAMBreakdown(weights_gb=6.0, optimizer_gb=12.0, gradients_gb=6.0, activations_gb=2.5381983657704925, overhead_gb=0.8)

```

The gate approves the run as a ~4 GB QLoRA job. What actually trains is a full fine-tune of 1.5B parameters. The predictor's full-FT figure assumes fp32 master weights and PRM loads bf16 on CUDA, but weights + grads + two Adam moments at 2 bytes each is already about 12 GB before activations. That is under-prediction, the unsafe direction: on Linux it ends in an OOM, and under Windows WDDM it spills to host memory without an error (the failure mode behind #649). (The gate is skipped for the default `batch_size: auto`, so this bites once a user pins the batch size.)

## Why it matters

- A user who leaves the default, or writes `quantization: 4bit` on purpose to fit a card, gets a full-precision load. For `distill` that means two models, including a frozen teacher that is usually the larger one.

- The stored config, and with it the registry entry and the config hash, records `4bit`, so the artefact misdescribes how it was trained.

- PRM ignoring `lora.*` means `lora.r` / `alpha` / `target_modules` in a PRM config are decorative, and the schema default `r: 64` suggests an adapter run that never happens.

This is the training-side counterpart of #367, which covers the eval side (`live_eval.load_model_and_tokenizer` judging an NF4 run on a bf16 base). They share a fix site (`unlearn.py` goes through `load_model_and_tokenizer`) but are separate defects.

## Proposed fix

For each trainer, pick one of two, deliberately:

1. **Honour it.** Build the quantisation config through the same Quant Menu path SFT uses (`utils/quant_menu.build_quantization_config_for_loader`, `sft.py:1331`), call `prepare_model_for_kbit_training` where the base is trainable, and pass `quantization=` into `load_model_and_tokenizer` for unlearn. PRM additionally needs a LoRA path (or `lora.r: 0` semantics) to be meaningful under 4-bit.

2. **Refuse it.** Add a schema cross-validator that rejects `quantization != 'none'` for tasks that cannot honour it, with a message naming the task. Because the default is `4bit`, this probably means resolving the default per task, so that a minimal config for these tasks still parses.

Either way, `_build_hardware_fit_input` should derive `peft` from what the trainer will actually do, not from the field alone.

## Acceptance criteria

- [ ] For each of the six tasks, a test either asserts a `quantization_config` reaches `from_pretrained` / `load_model_and_tokenizer`, or asserts the config is refused at parse time. The first kind must fail on current code.

- [ ] A test asserts PRM either applies `lora.*` or refuses a non-default LoRA block.

- [ ] A pre-flight test: `task: prm` with an int `batch_size` is not predicted as `qlora`.

- [ ] `docs/training.md` states per task which `quantization` values are honoured.

---

**Related, filed the same day:**

- #797 — training.gradient_checkpointing is decided by the trainer, not the config: `false` cannot turn it off on an...
- #805 — prm, moe_lora_routing, unlearn and ppo drop the optimizer and schedule: the default config's cosine/warmup/...

**Comment by Srinivasan8888** [comment]
I'd like to take this with option **2 (refuse)** for all six, shaped like your #983 ruling on #806. Honouring 4-bit per trainer is a separate feature for each, and several need a LoRA path first (PRM) or a k-bit prep decision (distill's frozen teacher). Measured on `main` (`5ee1278`) before claiming:

**Shipped configs this could break:** I loaded all 211 recipes, examples and templates. Only three use these tasks, the Whisper ASR recipes, and all three already set `quantization: none`. Two in-code generators emit `task: distill`: `soup draft` writes `quantization: none` explicitly (`commands/draft.py:390`), and `soup shrink` leaves it unset (`commands/shrink.py:498`).

**Plan:**
- **The default resolves per task.** For `distill`, `classifier`, `reranker`, `cross_encoder`, `prm`, `moe_lora_routing`, `unlearn` and `asr`, an **unset** `training.quantization` resolves to `none`, so a minimal config parses and the stored config, config hash and registry entry say what actually trained. An **explicit** non-`none` value is refused with a message naming the task. #321's note (schema.py:6005) says fields-set gates can trip on dump-and-reload. That doesn't bite here: a dumped config for these tasks carries the resolved `none`, and I checked that nothing re-validates stored registry dicts; `ship`, MCP and the UI reload user-written YAML.
- **PRM refuses an explicit `lora` block**, since `prm.py` never reads one. The schema default `r: 64` isn't treated as intent.
- **The pre-flight follows the resolved value**: `task: prm` with an int `batch_size` is budgeted as a full-precision run, not `qlora`.
- **`docs/training.md`** states per task which `quantization` values are honoured.

**One consequence to flag:** a `task: prm` config that never set `quantization` will now hash with `none` instead of `4bit`, so its `ship` config sha changes once. That's the point, since the old sha described a run that didn't happen, but it will look like a config change on the first rebuild.

**Overlap:** #868 (@Oxygen56, #805) edits `prm.py`, `mole_routing.py`, `unlearn.py` and `ppo.py`. This change stays in `schema.py`, `commands/train.py` and the docs, and touches no trainer file, so the two shouldn't conflict. CPU-only.

**Comment by MakazhanAlpamys** [comment]
Closed by #1043, thanks to @Srinivasan8888.

On the eight tasks that never apply quantization, `training.quantization` now resolves to `none` instead of being validated, documented and then ignored. An explicitly requested value is refused by name. A value Soup itself wrote into a config warns and resolves rather than failing the load — refusing that case is a next-release change, so no shipped config breaks today.

The ordering is pinned: moving the resolver later in the validator chain fails a dedicated test, so a future reordering cannot reopen this quietly.

Verified before merge on the tree merged against current `main`: 121 passed across the PR's tests and the two documentation guards, 783 passed / 8 skipped across the config blast radius, and all 48 shipped config dumps load as `none` with one warning each.
